src/backend/distributed/planner/deparse_shard_query.c

```diff
@@ -486,33 +501,80 @@ DeparseTaskQuery(Task *task, Query *query)
 
 
 /*
- * TaskQueryString generates task->queryStringLazy if missing.
+ * GetTaskQueryType returns the type of the task query.
+ */
+int
+GetTaskQueryType(Task *task)
+{
+	return task->taskQuery.queryType;
+}
+
+
+/*
+ * TaskQueryStringForAllPlacements generates task query string text if missing.
  *
  * For performance reasons, the queryString is generated lazily. For example
  * for local queries it is usually not needed to generate it, so this way we
  * can skip the expensive deparsing+parsing.
  */
 char *
-TaskQueryString(Task *task)
+TaskQueryStringForAllPlacements(Task *task)
 {
-	if (task->queryStringLazy != NULL)
+	if (GetTaskQueryType(task) == TASK_QUERY_TEXT_LIST)
+	{
+		return StringJoin(task->taskQuery.data.queryStringList, ';');
+	}
+	if (GetTaskQueryType(task) == TASK_QUERY_TEXT)
 	{
-		return task->queryStringLazy;
+		return task->taskQuery.data.queryStringLazy;
 	}
-	Assert(task->queryForLocalExecution != NULL);
+	Query *jobQueryReferenceForLazyDeparsing =
+		task->taskQuery.data.jobQueryReferenceForLazyDeparsing;
+	Assert(task->taskQuery.queryType == TASK_QUERY_OBJECT &&
+		   jobQueryReferenceForLazyDeparsing != NULL);
 
 
 	/*
-	 * Switch to the memory context of task->queryForLocalExecution before generating the query
+	 * Switch to the memory context of task->jobQueryReferenceForLazyDeparsing before generating the query
 	 * string. This way the query string is not freed in between multiple
 	 * executions of a prepared statement. Except when UpdateTaskQueryString is
-	 * used to set task->queryForLocalExecution, in that case it is freed but it will be set to
+	 * used to set task->jobQueryReferenceForLazyDeparsing, in that case it is freed but it will be set to
 	 * NULL on the next execution of the query because UpdateTaskQueryString
 	 * does that.
 	 */
 	MemoryContext previousContext = MemoryContextSwitchTo(GetMemoryChunkContext(
-															  task->queryForLocalExecution));
-	task->queryStringLazy = DeparseTaskQuery(task, task->queryForLocalExecution);
+															  jobQueryReferenceForLazyDeparsing));
```

Comment:
I'm not 100% sure I buy the correctness of this, but that's not this PR's fault.